### PR TITLE
[AGENT-3690] Refactor DrumServerRun and improve shutdown

### PR DESCRIPTION
**Pull request by elatt**
# This repository is public. Do not put here any private DataRobot or customer's data: code, datasets, model artifacts, .etc.

## Summary
- Adds logging
- Move shutdown logic into a single method
- Improve shutdown logic to always be sure to kill the processes
- No need to import 3rd party `psutil` library
- Use `TimeoutError` instead of `assert`

## Rationale
When debugging some issues with a new test, I found working with this class to be hard to debug with as it was lacking enough logging and had incorrect shutdown logic for some edge cases.

### CODE REVIEW

**Comment by yakov-g** [comment]
jenkins run test_integration_general test_integration_r_environment

**Comment by elatt** [comment]
I think I may have introduced real bugs... when glancing through the failures, it looks like some of the tests weren't shutting down the drum server fast enough (which is a new error we are detecting now), and one of the R servers were starting up too slowly (which shouldn't be a change from the old code... so maybe that was just a flake)

**Comment by elatt** [comment]
Jenkins run test_integration_general

**Comment by elatt** [comment]
jenkins run test_integration_general test_integration_r_environment

**Comment by elatt** [comment]
jenkins run test_integration_general

**Comment by elatt** [comment]
jenkins run test_integration_r_environment

**Comment by elatt** [comment]
jenkins run test_integration_general

**Comment by yakov-g** [comment]
jenkins run test_integration_general test_integration_r_environment

**Comment by elatt** [comment]
jenkins run test_integration_general

File: model_templates/flask_extension_httpauth/README.md [hunk 1]
```diff
@@ -11,7 +11,9 @@ Note: it is **not** necessary (nor recommended) to add authentication to custom
 This example is simply to demonstration the flexibility of the `custom_flask.py` hook.
 
 ## Instructions
-Create a new custom model with these files and use the Python Drop-In Environment with it
+Create a new custom model with these files and use the Python Drop-In Environment with it.
+
+**Important:** extending the web server is only available when running **without** the `--production` flag (or `PRODUCTION=1` environment variable).
```
Comment (elatt) [inline model_templates/flask_extension_httpauth/README.md:16]: @yakov-g is this sufficient for noting that the new feature doesn't work in production mode? I took a look to see if we can have the actual application issue a warning if it found a `custom_flask.py` file in the model-dir **and** `--production` was active but I couldn't see a good way to do that so I'll leave that as a future exercise.

Comment (yakov-g) [reply model_templates/flask_extension_httpauth/README.md:16]: Yes, looks good.

**Comment by elatt** [comment]
jenkins run test_integration_r_environment

**Comment by elatt** [comment]
jenkins run test_integration_general

**Comment by elatt** [comment]
Jenkins run test_integration_python_onnx

**Comment by yakov-g** [comment]
hooray!

**Review by scottp-ml — APPROVED** [review]
LGTM
